### ISSUE TITLE
Deny Public Access Redirect Hook

### ISSUE THREAD

**Issue by ericp-mrel** [wpengine/headless-framework#389]
Looking at the `wpe_headless_deny_public_access` hook, this seems needlessly complex and prone to constant maintenance in order to account for lots of possible plugin conflicts and so forth. Is there a reason the `template_redirect` action wasn't used instead since I believe that is only fired for template related routes and would therefore only be fired to front-end related requests and would avoid needing to check for GraphQL, REST, etc routes and having to constantly adjust the checks if there are future conflicts.

https://developer.wordpress.org/reference/hooks/template_redirect/

Mainly just curious what the reasoning was for choosing the `parse_request` action was instead of using the `template_redirect` hook, because the `template_redirect` hook seems to work perfectly for this specific situation.

**Comment by blakewilson** [comment]
Hey @ericp-mrel,

Thanks for your question.

It looks like we're using [`parse_request`](https://developer.wordpress.org/reference/hooks/parse_request) instead of [`template_redirect`](https://developer.wordpress.org/reference/hooks/template_redirect/) to get access to the [`WP`](https://developer.wordpress.org/reference/hooks/parse_request/#parameters) parameter, which includes the request object.

This is something `template_redirect` does not provide, and is needed to assemble the [redirect url](https://github.com/wpengine/headless-framework/blob/canary/plugins/wpe-headless/includes/deny-public-access/callbacks.php#L40).

**Comment by ericp-mrel** [comment]
Couldn't you use the global `$wp` object to get access to the WP parameter so you wouldn't necessarily need it to be part of the hook's parameters. I actually had to something similar in my own plugin and redirect front-end requests to a different domain and it seems to work for me using the `template_redirect` action with the global `$wp` object.

Hopefully this would work for this plugin as I believe it would make it much easier to maintain in the long term.

**Comment by mindctrl** [comment]
I don't remember why `parse_request` was chosen. Pinging @rfmeier and @nickcernis to see if they do.

Looking at the hooks that run between `parse_request` and `template_redirect` makes me wonder if it was performance related. According to Query Monitor, on a simple site with very few plugins, the following hooks run between the two:
- `send_headers`
- `parse_tax_query`
- `parse_query`
- `pre_get_posts`
- `posts_selection`
- `parse_term_query`
- `pre_get_terms`
- `metadata_lazyloader_queued_objects`
- `wp`

I think it would be more with other plugins. WooCommerce hooks into `parse_query`, `pre_get_posts`, and `wp` and runs queries, sets other filters, etc. A lot of other plugins use these hooks. A number of WP core functions hook into `template_redirect`.

Some functions hooked into those run other actions and filters, like how WP hooks into `send_headers` and runs `nocache_headers()`, which then runs a `nocache_headers` filter, which plugins tie into.

Basically it opens up the door for a lot more code to run before redirecting. Do we need that code to run? I don't know for sure. Do we know of use cases that require us to adjust the existing logic to fix a conflict? Do we know of reasons for that code to run in the context this plugin runs with this feature enabled?

**Comment by nickcernis** [comment]
I think we hooked as early as we could for performance and to reduce the chance of other code redirecting/altering the request first.

I'm good with `template_redirect` if we think it will improve maintainability without impacting speed or functionality.
